### DISCUSSION TITLE
Multi-root workspace support ?

### DISCUSSION THREAD

**Discussion by lljr** [joaotavora/eglot#909] [Q&A] [answered]
https://github.com/Microsoft/vscode-extension-samples/tree/main/basic-multi-root-samples multi-root support. Now, https://github.com/joaotavora/eglot/discussions/778#discussioncomment-1877274 shows an example of trying to set this up. I have had no success doing this for  a React project. Given my very limited Elisp knowledge, does #893 solve multi-root support? If so, is there an easy way to set this up? I've failed to find an example in the repo that teaches how to do so. 

Thanks in advance :).

**Comment by joaotavora** [comment]
Hi @lljr. It's _possible_ that #893 indeed solves it.  But you are right, we are missing an example in the README.md that demonstrates how it is used.  If a simple example can derive from this discussion it would be great.

I'm not a user of multi-root projects myself, but my understanding is that if you setup an Emacs `project.el` project `x` so that:

* `(project-root x)` returns the "primary" root 
*  `(project-external-roots x)` returns a list of the secondary(?) roots

... then invoking `M-x eglot` inside that project ("inside" meaning `project-current` will return `x`) will cause the primary and secondary root information will be sent to the LSP server:

* in the `workspaceFolders` argument of the `initialize` client-to-server request
* as the response to the `workspace/workspaceFolders` server-to-client request

@dgutov is the `project.el` specialist and maintainer. @nemethf also has some knowledge in this deparment, I gather.

**Reply by nemethf** [reply]
I'm not aware of any user-level documentation of project.el, but from the NEWS.md:

    Eglot now advertises `project-root` and `project-external-roots` as
    workspace-folders.  (Configuring `tags-table-list` sets the external
    roots of a simple git project.)

**Reply by joaotavora** [reply]
Right, I forgot about that.  That's minimal documentation, but the sentence in parenthesis is very mysterious :-)

**Reply by lljr** [reply]
Hi @nemethf  @joaotavora . Should I see in the `:workspaceFolders` arg, in the client-to-server request, the same paths I added to `tags-table-list`? If so, that's not the case... I only see the path for `project-root` and not the ones I added to `tags-table-list`. 

EDIT: However, this works, at least for [webpack resolved modules ](https://webpack.js.org/configuration/resolve/#resolvealias). After I add the "sub-root" folders as said in 

```
Eglot now advertises `project-root` and `project-external-roots` as
workspace-folders.  (Configuring `tags-table-list` sets the external
roots of a simple git project.)
```

specifically, `tags-table-list`. I get auto completion from the configured webpack modules, not just the relative path for the imported variable or function. When I start Emacs _without_ setting the "sub-root" directories via `tags-table-list` I *don't* get auto-completion for webpack resolved modules. I get completion to the variable or function in it's relative path. This is due to the way my repo is setup with Next.js, which you can get a clearer idea by quickly reading: https://stackoverflow.com/a/61838859.

That said, perhaps it's a bug that I don't see the extra directories in `:workspaceFolders` arg in the client-server request? 

Now, it would be nice that this could all be done automagically :). Maybe through some user-defined setter that binds a string  `file.extension` to a variable and tells Eglot to set its directory as "sub-root" of a given project and adds them to the `:workspaceFolders` arg.

EDIT 2: I guess it may not be a bug. I launched the elisp debugger and called eglot (`M-x eglot`) and when it stopped at https://github.com/joaotavora/eglot/pull/893/commits/01699c8d2c6d5659a11bb053a538f7cc72bc7f9b#diff-d6c5a55ddb4dd8dca075c4e0d48c851460ead14c77f333c66b02ccc7d1c2aff0R727, the value of `dirs` is the same value I see in `:workspaceFolders` arg in the client-server request. To reiterate, this is _after_ I set `tags-table-list` to the "sub-root" folders I need. Keep in mind I double checked by eval'ing `tags-table-list` and I do see the paths  I need in `tags-table-list`. 

To all this, it's still a bit confusing how all these vars work because in `project.el` the docstring for `project-vc-external-roots-function` says 

```
"Function that returns a list of external roots.

It should return a list of directory roots that contain source
files related to the current buffer.

The directory names should be absolute.  Used in the VC project
backend implementation of `project-external-roots'."
```

this talks about directory roots but the docstring for `tags-table-list` says

```
Documentation:
List of file names of tags tables to search.
An element that is a directory means the file "TAGS" in that directory.
```

which imply that `tags-table-list` should be set to the absolute path including the file, and if no element with file is included, means that "TAGS" is in the directory, which is not necessarily in this case (it's `jsconfig.json`).

**Comment by lljr** [comment]
As @nemethf, adding the "sub-root" dirs to `tags-table-list` solves the problem. Hopefully more is added to this feature in the future.

**Reply by joaotavora** [reply]
OK, good. Can you explain exactly what you did in terms of your `.emacs` and `tags-table-list` to add "subroots".  I'd like to put this information in the README.md. A Pull Request to that effect is also welcome :-)

> Hopefully more is added to this feature in the future.

What would you like to see added?

**Reply by lljr** [reply]
@joaotavora this stopped working with the latest commit in master. I use to be able to do `(setq tags-table-list
      '("~/src/proj/path2/file.json" "~/src/proj/path1/file.json))` and I would see the paths in `workspaceFolders` server-client response. Has something changed since then? 

What I think should be added is  a way to set the files in a variable that indicate another "workspaceFolder". For example, in Typescript lang server the existance of a `jsconfig.json` or `tsconfig.json` inside a directory below the root indicate a "sub-root".

**Reply by joaotavora** [reply]
Please write a full recipe starting with `emacs -Q` that explains exactly what you want to do. A full but minimal recipe, containing the server used, a simple project, and all the things that you do. Only then can I help you. For example I don't understand what you mean by "see the paths in workspaceFolders server-client response."

Also, what "latest commit in master" are you referring to? The latest commit is about `clojure-lsp`: I doubt that has any bearing to your problem, which seems to be about workspaces?  If you could bisect and come up with the actual commit where this stopped working it would be helpful. (and verify that it is indeed worked with the commit before).

**Reply by joaotavora** [reply]
Perhaps pinging @nemethf can help as he added this functionality.  I can't think of any Eglot commit that has touched this code.

**Reply by lljr** [reply]
@joaotavora thaks for the quick response. I will get to you later

**Reply by nemethf** [reply]
> Perhaps pinging @nemethf can help as he added this functionality. 

Maybe I can help when I see the full recipe.  Nevertheless, it's worth checking the return value of `project-external-roots`.

**Comment by quotuva** [comment, marked as answer]
FWIW, for `jdtls` I use:

    (push `((java-mode java-ts-mode) "jdtls" "--jvm-arg=-javaagent:/home/user/lombok.jar"
    	:initializationOptions
    	,(lambda (server)
    	   `(:workspaceFolders ,(vconcat (mapcar (lambda (o) (plist-get o :uri)) (eglot-workspace-folders server))))))
          eglot-server-programs)


~and comment [this line](https://github.com/joaotavora/eglot/blob/master/eglot.el#L1585) in `eglot.el`, to work around https://github.com/eclipse-jdtls/eclipse.jdt.ls/issues/1245~

EDIT (by @joaotavora):  Commenting this line in `eglot.el` is not needed apparently.

EDIT2: This is indeed down to a bug in the `jdtls` server.  Go here to support its fix https://github.com/eclipse-jdtls/eclipse.jdt.ls/issues/1245 (I'd say it shouldn't be terribly hard).

**Reply by joaotavora** [reply]
Seems like that `vconcat` is simply getting a vector from a list.  Maybe `eglot-workspace-folders` should be retuning a vector instead?  Or maybe putting the vconcat trick in the line you commented (instead of commenting it) would work as well?  Can you try and report back?

**Reply by quotuva** [reply]
Oh, it seems that the commenting is not required (it was a vestige of my attempts to fix it). But that issue I linked to above is related to `jdtls`'s odd way of using `:workspaceFolders`: at initialization, and (as I've discovered) without the `:uri`, `:name` etc. I don't think there's anything to fix in `eglot`.

**Reply by joaotavora** [reply]
OK. Also, I've confirmed that `eglot-workspace-folders` already returns a vector.  So are you sure your hand-rolled `vconcat` is needed?

**Reply by quotuva** [reply]
Yes, I checked again; probably because of `mapcar` returning a list:  `json-serialize (wrong-type-argument symbolp "file:///first item in tags-table-list")`.

**Reply by joaotavora** [reply]
But what is the mapcar doing there anyway?

**Reply by joaotavora** [reply]
Maybe is it because `jdtls` wants just a vector of strings in that place? Then replace the `elt` with `(plist-get i :uri)`, I'd say.

Has jdtls been made aware of this problem?

**Reply by quotuva** [reply]
> `jdtls` wants just a vector of strings in that place?

Exactly.
> Then replace the `elt` with `(plist-get i :uri)`

Yes, that's better.
> Has jdtls been made aware of this problem?

Yes, the issue I'd mentioned above; it also has a link to the code.
